src/providers/wms/qgswmsprovider.cpp

```diff
@@ -215,6 +215,36 @@ QString QgsWmsProvider::getTileUrl() const
   }
 }
 
+static bool isValidLegend( const QgsWmsLegendUrlProperty &l )
+{
+  return l.format.startsWith( "image/" );
+}
+
+/**
+ * Picks a usable legend URL for a given style.
+ */
+static QString pickLegend( const QgsWmsStyleProperty &s )
+{
+  QString url;
+  for ( int k = 0; k < s.legendUrl.size() && url.isEmpty(); k++ )
+  {
+    const QgsWmsLegendUrlProperty &l = s.legendUrl[k];
+    if ( isValidLegend( l ) )
+    {
+      url = l.onlineResource.xlinkHref;
+    }
+  }
+  return url;
+}
+
+static const QgsWmsStyleProperty *searchStyle( const QVector<QgsWmsStyleProperty> styles, const QString& name )
+{
+  foreach ( const QgsWmsStyleProperty &s, styles )
+    if ( s.name == name )
+      return &s;
```

Comment:
This returns a pointer to an element of a temporary copy of the vector. It may work because of implicit sharing but it looks rather fragile.

Better

  * take a const reference as parameter
  * return a copy